core/state_connector_test.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/flare-foundation/coreth/core/vm"
+)
+
+type mockMessage struct {
+	from       common.Address
+	to         common.Address
+	gas        uint64
+	gasPrice   *big.Int
+	gasFeeCap  *big.Int
+	gasTipCap  *big.Int
+	value      *big.Int
+	data       []byte
+	accessList types.AccessList
+}
+
+func (m *mockMessage) From() common.Address         { return m.from }
+func (m *mockMessage) To() *common.Address          { return &m.to }
+func (m *mockMessage) GasPrice() *big.Int           { return m.gasPrice }
+func (m *mockMessage) GasFeeCap() *big.Int          { return m.gasFeeCap }
+func (m *mockMessage) GasTipCap() *big.Int          { return m.gasTipCap }
+func (m *mockMessage) Value() *big.Int              { return m.value }
+func (m *mockMessage) Gas() uint64                  { return m.gas }
+func (m *mockMessage) Nonce() uint64                { return 0 }
+func (m *mockMessage) Data() []byte                 { return m.data }
+func (m *mockMessage) AccessList() types.AccessList { return m.accessList }
+func (m *mockMessage) IsFake() bool                 { return true }
+
+type mockStateConnectorCaller struct {
+	context  vm.BlockContext
+	testCase string
+	msg      *mockMessage
+}
+
+func (m *mockStateConnectorCaller) Call(caller vm.ContractRef, addr common.Address, input []byte, gas uint64,
+	value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+	switch {
+	case m.testCase == "happy_path" || m.testCase == "finalized_data_error":
+		voter1 := common.BytesToHash([]byte("voter1"))
+		voter2 := common.BytesToHash([]byte("voter2"))
+		merkleRootHash := []byte("some_hash")
+
+		if caller == vm.AccountRef(m.msg.From()) {
+			if addr == common.HexToAddress(GetPrioritisedFTSOContract(nil)) {
+				return []byte("voter"), 0, nil
+			}
+			if addr == common.BytesToAddress([]byte("voter")) {
+				return append([]byte{}, append(voter1.Bytes(), voter2.Bytes()...)...), 0, nil
+			}
+		}
+		if caller == vm.AccountRef(common.BytesToAddress(voter1.Bytes())) ||
+			caller == vm.AccountRef(common.BytesToAddress(voter2.Bytes())) {
+			return merkleRootHash, 0, nil
+		}
+		if caller == vm.AccountRef(stateConnectorCoinbaseSignalAddr(nil, nil)) {
+			if m.context.Coinbase != stateConnectorCoinbaseSignalAddr(nil, nil) {
+				return nil, 0, fmt.Errorf("invalid coinbase address: expected %v, got %v", stateConnectorCoinbaseSignalAddr(nil, nil), m.context.Coinbase)
+			}
+			if m.testCase == "happy_path" {
+				return nil, 0, nil
+			}
+			return nil, 0, errors.New("finalization error")
+		}
+	case m.testCase == "default_attestors_error_1":
+		return nil, 0, errors.New("voter whitelister error")
+	case m.testCase == "default_attestors_error_2":
+		switch addr {
+		case common.HexToAddress(GetPrioritisedFTSOContract(nil)):
+			return []byte("voter"), 0, nil
+		case common.BytesToAddress([]byte("voter")):
+			return nil, 0, errors.New("list price provider error")
+		}
+	}
+	return nil, 0, errors.New("undefined test case")
+}
```

Comment:
I really don't like this approach. IMO we should not rely on a string that describes the test to trigger a different mechanism. Instead, the [mock should have a function as its attribute, which can be overloaded by each test](https://optakt.github.io/code/testing/#mocks).

This way, the mock itself is simple, and all of the complexity that describes the test functionally and _specifies_ it so to speak, is contain within the test itself.